dbms/src/DataTypes/DataTypeFactory.cpp

```diff
@@ -40,22 +40,44 @@ DataTypePtr DataTypeFactory::get(const String & full_name) const
     ASTPtr ast = parseQuery(parser, full_name.data(), full_name.data() + full_name.size(), "data type", 0);
     return get(ast);
 }
+// DataTypeFactory is a Singleton, so need to be protected by lock.
+DataTypePtr DataTypeFactory::getOrSet(const String & full_name)
+{
+    {
+        std::shared_lock lock(rw_lock);
+        auto it = fullname_types.find(full_name);
+        if (it != fullname_types.end())
+        {
+            return it->second;
+        }
+    }
+    ParserIdentifierWithOptionalParameters parser;
+    ASTPtr ast = parseQuery(parser, full_name.data(), full_name.data() + full_name.size(), "data type", 0);
+    DataTypePtr datatype_ptr = get(ast);
+    // avoid big hashmap in rare cases.
+    if (fullname_types.size() < MAX_FULLNAME_TYPES)
```

Comment:
> `full_name` is the name of data type. Can the number of `full_name` greater than 10000?

The number of `full_name` is less than 100 (including the [nullable](https://github.com/pingcap/tiflash/blob/d8c369cb8979efeda8df9ef6b67dfd9c38a7d9c9/dbms/src/Storages/Transaction/TypeMapping.cpp#L207) prefix datatype).
![image](https://user-images.githubusercontent.com/14118780/206496080-85fe0424-91cc-4398-88df-463ba80e846b.png)
But if the user create tables by clickhouse-client, the `full_name` can combine [these datatypes](https://github.com/pingcap/tiflash/blob/d8c369cb8979efeda8df9ef6b67dfd9c38a7d9c9/dbms/src/Storages/Transaction/TypeMapping.cpp#L192) with [tuple](https://clickhouse.com/docs/en/sql-reference/data-types/tuple), it may generate too many `full_name`, so I add this check to avoid big hashmap.

Reply:
if we support to store `varchar(200)` as `FixedString(200)` in the future, it may generated different full_name? Though we don't support it now.

Reply:
> if we support to store `varchar(200)` as `FixedString(200)` in the future, it may generated many different full_name? Though we don't support it now.

It seems that [`varchar(200)` is not a `FixedString`, char(0) ~ char(255)  is a `FixedString`](https://dev.mysql.com/doc/refman/5.7/en/char.html), so it will not generate too many different `full_name`.